### DISCUSSION TITLE
Lesson 4: Web3.py Simple Storage errors

### DISCUSSION THREAD

**Discussion by betheashvin** [smartcontractkit/full-blockchain-solidity-course-py#1542] [Q&A] [answered]
Issue 1: What is the issue here?: When i try to import web3, solcx, & dotenv, it gives me an error saying: Import "web3" could not be resolved Import "solcx" could not be resolved Import "dotenv" could not be resolved. I even tried copy pasting the full code to ensure there isn't any error in the code, but i guess there's something else out of the way. I tried what internet suggested me, but so far did not prove to be helpful. So can someone please be kind enough to help me out here? Thank you. 
When i do pip install web3, i get error :
...  C:\Users\ashvi\AppData\Local\Programs\Python\Python39\include\pyconfig.h(59): fatal error C1083: Cannot open include file: 'io.h': No such file or directory
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.32.31326\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> lru-dict

note: This is an issue with the package mentioned above, not pip.
hint: See above for output from the failure.

And when i run the program code, the detailed Terminal Output appears as:
Traceback (most recent call last):
  File "c:\Users\ashvi\demos\web3_py_simple_storage\deploy.py", line 1, in <module>
    from solcx import compile_standard, install_solc
ModuleNotFoundError: No module named 'solcx'

Code link: https://github.com/PatrickAlphaC/web3_py_simple_storage/blob/main/deploy.py

Python version: I had the latest one - version 10 but changed it to version 9 as i heard v10 has some issues with web3 lib.
Interpreter: using the recommended one by vscode.

Issue 2: Should i put ...python/script in path (system) variable or node.js location? - When i alter the path in system variable and change it to python, vscode does not recognize node.js, while when i enter node.js as path, VSCode has some issues with python, i reckon, as i remove "...python/script" file from path variable in system environment variables. This is as when i do node --version, the terminal output is:

node : The term 'node' is not recognized as the name of a cmdlet, function, script file, 
or operable program. Check the spelling of the name, or if a path was 
included, verify that the path is correct and try again.
At line:1 char:1
+ node --version
+ ~~~~
+ CategoryInfo          : ObjectNotFound: (node:String) [], CommandNotFoundException
+ FullyQualifiedErrorId : CommandNotFoundException

I'm confused, someone please help me resolve these issues. Much appreciated.

**Comment by cromewar** [comment, marked as answer]
Hello @betheashvin 

> Issue 1: What is the issue here?: When i try to import web3, solcx, & dotenv, it gives me an error

This is because you are not selecting the correct python interpreter on which you have those packages installed, make sure to select the correct python interpreter on python.

> .. C:\Users\ashvi\AppData\Local\Programs\Python\Python39\include\pyconfig.h(59): fatal error C1083: Cannot open include file: 'io.h': No such file or directory

This is because you have some important MSVC libraries needed for python development, for this you would have to install visual studio community and the python libraries of it, this is very heavy so you will need at least 35Gb of free space.

> ssue 2: Should i put ...python/script in path (system) variable or node.js location? 

Neither of them, you should add global paths for both Node and Python in order to work anywhere you wanted to work.

Recommendation:

Windows is just really terrible managing this things, I strongly recommend you to use WSL (Windows Subsystem for Linux) and migrate your workflow there, there is a guide on the course index called "Windows Support" to install WSL and and the tools you'll need.

**Reply by betheashvin** [reply]
Thank you. I'll work on this.

**Reply by betheashvin** [reply]
Hello. I actually had installed the VS Installer and the python libraries and others from VS Community, but it gives the same error as above. I tried all the available python interpreters in VSCode but it doesn't seem to help. Should i uninstall and reinstall python version 3.9?
Thanks.

**Reply by cromewar** [reply]
you may have missing some dependencies for development on your windows machine.
I recommend python 3.8 instead, and again WSL is a better option.

**Comment by betheashvin** [comment]
You mean using WSL instead of the Windows Terminal and using VSCode as it
is, right? Or something else?
And sorry I'm just confused in this.
Thanks again :)

**Reply by cromewar** [reply]
WSL is a instance of linux running on your windows machine, please give a check to [this](https://medium.com/@cromewar/how-to-setup-windows-10-11-for-smart-contract-development-and-brownie-e7d8d13555b3) guide.

**Reply by betheashvin** [reply]
Hey. I did try using Ubuntu/WSL, python 3.8.10, VS Community/Build Tools, but it still didn't help. Any recommendations as to how i may proceed further? 
Thanks.
Regards.

**Reply by cromewar** [reply]
If you have installed WSL? could you please share the steps you followed to execute de code?
Because as WSL does no depends on any microsoft external library.

If you find this problem unsolvable, please contact me trough email and we can set up a meeting.

Cheers.

**Reply by betheashvin** [reply]
I reckon the issue is now resolved, grateful for your help. But i ran into two others. The base one i believe is when i face trouble when running "nvm install node" code on ubuntu / vscode(wsl terminal.) Please do check these attached images. FYI: These error lines seem never ending.

![Screenshot (1076)](https://user-images.githubusercontent.com/68969607/174883611-aa4e5570-643e-41a7-897a-5991f56fd383.png)
![Screenshot (1072)](https://user-images.githubusercontent.com/68969607/174883684-382e47fe-7339-45dc-88c2-dedcacc206c7.png)

![Screenshot (1081)](https://user-images.githubusercontent.com/68969607/174883712-dcd31110-65d4-48bc-b019-fe2a9b62ae01.png)
![Screenshot (1079)](https://user-images.githubusercontent.com/68969607/174883702-e7f6fc69-4137-4bf7-b331-ddaf40d01f6f.png)

Thanks.

**Comment by cromewar** [comment]
Hello @betheashvin 
Your WSL needs GCC, so please:
`sudo apt install gcc`

And, also based on the compatibility with NPM and web3 Libraries please uninstall the current version and install the LTS one:
``` Shell
nvm deactivate
nvm uninstall  // you might select the version you want to uninstall
nvm install --lts
```

**Reply by betheashvin** [reply]
I checked with the gcc commands, it was already installed:

```
ashvin1@LAPTOP-THTH0RN2:~$ gcc --version gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0 
Copyright (C) 2019 Free Software Foundation, Inc. This is free software; see the 
source for copying conditions. There is NO warranty; not even for MERCHANTABILITY 
or FITNESS FOR A PARTICULAR PURPOSE.
```

Additionally the Ubuntu terminal is unable to run the code nvm deactivate or nvm uninstall 18.4.0. It says:

```
ashvin1@LAPTOP-THTH0RN2:~$ nvm deactivate Could not find /home/ashvin1/.nvm//bin in ${PATH} 
ashvin1@LAPTOP-THTH0RN2:~$ nvm deactivate 18.4.0 Could not find /home/ashvin1/.nvm//bin in ${PATH} 
ashvin1@LAPTOP-THTH0RN2:~$ nvm uninstall 18.4.0 N/A version is not installed...
```

When i run "nvm ls" code, i get:
```
ashvin1@LAPTOP-THTH0RN2:~$ nvm ls 
v16.15.1 
v18.4.0 
-> system 
iojs -> N/A (default) 
unstable -> N/A (default)
node -> stable (-> v18.4.0) (default) 
stable -> 18.4 (-> v18.4.0) (default) 
lts/* -> lts/gallium (-> N/A) 
lts/argon -> v4.9.1 (-> N/A) 
lts/boron -> v6.17.1 (-> N/A) 
lts/carbon -> v8.17.0 (-> N/A) 
lts/dubnium -> v10.24.1 (-> N/A) 
lts/erbium -> v12.22.12 (-> N/A) 
lts/fermium -> v14.19.3 (-> N/A) 
lts/gallium -> v16.15.1 (-> N/A)
```
The output of code "nvm debug" is:
```
ashvin1@LAPTOP-THTH0RN2:~$ nvm debug
nvm --version: v0.38.0
$SHELL: /bin/bash
$SHLVL: 1
whoami: 'ashvin1'
${HOME}: /home/ashvin1
${NVM_DIR}: '${HOME}/.nvm'
${PATH}: ${HOME}/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Program Files/Microsoft SQL Server/150/Tools/Binn/:/mnt/c/Program Files/Git/cmd:/mnt/c/Program Files/Microsoft SQL Server/Client SDK/ODBC/170/Tools/Binn/:/mnt/c/Program Files/dotnet/:/mnt/c/Program Files (x86)/IncrediBuild:/mnt/c/ProgramData/chocolatey/bin:/mnt/c/Users/ashvi/AppData/Local/Programs/Python/Python38/Scripts/:/mnt/c/Users/ashvi/AppData/Local/Programs/Python/Python38/:/mnt/c/Users/ashvi/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/ashvi/.dotnet/tools:/mnt/c/Users/ashvi/AppData/Local/Programs/Microsoft VS Code/bin:/snap/bin
$PREFIX: ''
${NPM_CONFIG_PREFIX}: ''
$NVM_NODEJS_ORG_MIRROR: ''
$NVM_IOJS_ORG_MIRROR: ''
shell version: 'GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)'
uname -a: 'Linux 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux'
checksum binary: 'sha256sum'
OS version: Ubuntu 20.04.4 LTS
curl: /usr/bin/curl, curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
wget: /usr/bin/wget, GNU Wget 1.20.3 built on linux-gnu.
git: /usr/bin/git, git version 2.25.1
grep: /usr/bin/grep (grep --color=auto), grep (GNU grep) 3.4
awk: /usr/bin/awk, GNU Awk 5.0.1, API: 2.0 (GNU MPFR 4.0.2, GNU MP 6.2.0)
sed: /usr/bin/sed, sed (GNU sed) 4.7
cut: /usr/bin/cut, cut (GNU coreutils) 8.30
basename: /usr/bin/basename, basename (GNU coreutils) 8.30
rm: /usr/bin/rm, rm (GNU coreutils) 8.30
mkdir: /usr/bin/mkdir, mkdir (GNU coreutils) 8.30
xargs: /usr/bin/xargs, xargs (GNU findutils) 4.7.0
nvm current: system
which node: /usr/bin/node
which iojs:
which npm:
npm config get prefix:
Command 'npm' not found, but can be installed with:

sudo apt install npm
npm root -g:
Command 'npm' not found, but can be installed with:

sudo apt install npm
```

Using Ubuntu terminal version 2004.2022.1.0 & using WSL terminal on VSCode as well.

Kindly assist me in this. 
Thank you very much for your time.
Regards.

**Reply by cromewar** [reply]
Mmmm this is because the node installation is conflicting with your local node installation on windows, in this case I'll suggest to uninstall node from windows and also delete this folder 
`c/Users/ashvi/AppData/Roaming/npm`

this might help, but I've never seen this error before so I'll make more research on this.

**Reply by betheashvin** [reply]
I did the above steps then had these codes run:
sudo apt-get install curl
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.38.0/install.sh | bash
nvm install --lts

Unfortunately, it gave a failure output again. 
If you find sth helpful on this, please lmk. Difficult to proceed ahead.
Thanks.
Regards.

**Reply by betheashvin** [reply]
Hey @cromewar, someone suggested the below solution and it worked:
"Try renaming the ~/.nvm directory with mv ~/.nvm/ ~/.nvm.old. You may want to exit WSL first and wsl --terminate <distroname> if files are in use. Then try nvm install --lts again. Any difference? – 
[NotTheDr01ds]"
Just wanted to let you know. 
Regards.

**Reply by cromewar** [reply]
I'm glad is solved now!
